### PR TITLE
Decouple Sign task from PgpSignatory implementation

### PR DESCRIPTION
**Pull request by emlun**
**TL;DR:** Removing a runtime cast from `Signatory` to `PgpSignatory` allows plugins to provide arbitrary signing backends, with minimal or no change in behaviour.

### Background
The Signing plugin works great, except that requiring both secret keys and passphrases in files on disk is frankly a dealbreaker. It also doesn't help that GnuPG is getting rid of the secret keyring file outright. An alternative `Signatory` implementation could allow using an external `gpg` process as the backend, leveraging all of GnuPG's features including pinentry and support for keys on smartcards. This is in fact [how the maven-gpg-plugin works](https://svn.apache.org/viewvc/maven/plugins/tags/maven-gpg-plugin-1.6/src/main/java/org/apache/maven/plugin/gpg/GpgSigner.java?view=markup#l51).

### Problem
Implementing this support in a Gradle plugin turned out to be as easy as setting `project.signing.signatories = new CustomSignatoryProvider()` and have my own `Signatory` implementation call `gpg` as an external process. However at runtime the `Sign` task tries to cast whatever `Signatory` instance it receives to a `PgpSignatory`, which of course does not work with any other implementation.

### Solution
Fortunately, the only reason the `Sign` task casts to `PgpSignatory` is to call `PgpSignatory.getKeyId()` and use the result as a task input property. The input property is never read, so I assume it's used to allow Gradle's incremental builds to detect if the signing key changes. As such, any unique identifier for the key should work - and in fact `Signatory.getName()` is already specified to satisfy this requirement.

### Proposed changes
Get rid of the cast and use `Signatory.getName()` instead of `PgpSignatory.getKeyId().getAsHex()`. This is technically a change in behaviour, but its impact should be minimal and would not persist across `clean`s. It could however be worth considering changing the `getName()` specification to say it MUST be unique instead of SHOULD, or alternatively to add a `getId()` method or the like to the `Signatory` interface. That would more clearly communicate that it represents a machine-readable unique identifier and not a human-friendly somewhat-unique description.

Gradle Developers discussion: https://groups.google.com/forum/#!topic/gradle-dev/FINVhPpxW-c

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [ ] ~~[Link to Design Spec](https://github.com/gradle/gradle/tree/master/design-docs) for changes that affect more than 1 public API (that is, not in an `internal` package) or updates to > 20 files~~
- [ ] ~~Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective~~ *No tests added since no new feature was added*
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] ~~Update User Guide, DSL Reference, and Javadoc for public-facing changes~~
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by emlun** [comment]
There is now an alpha [gradle-gpg-signign-plugin][gpg-signing] available which integrates with this patch to support signing with recent GPG releases.

[gpg-signing]: https://github.com/Yubico/gradle-gpg-signing-plugin

**Comment by Minecrell** [comment]
There is also https://github.com/gradle/gradle/pull/1703 which is quite similar (it includes a slightly different fix for the same issue but it also directly adds a GPG agent signatory to Gradle).

**Comment by bmuschko** [comment]
This change is somewhat overlapping with https://github.com/gradle/gradle/pull/1703. I just reviewed https://github.com/gradle/gradle/pull/1703 and am awaiting comments from the contributor. If we happen to merge the pull request then we'll likely close this one as it will cover the same functionality.

**Comment by emlun** [comment]
Sorry I've been slow to respond, I've been traveling the last couple of days.

1703 is indeed very similar, funny that I didn't find it when I checked to see if someone had already taken this on. That is also in many ways a better patch. Mine seems less intrusive but also much less comprehensive, in particular adding the `getInputProperty` method to `Signatory` is something I would have liked to do but I didn't feel comfortable requesting a breaking API change. Perhaps some of my tests could be useful for 1703, though. I'll take a look at that.

I won't mind at all if this one is closed in favor of 1703, or however you choose to go at it, I'll be happy as long as we get a way to use GnuPG for signing. :) Please let me know if there's anything else I can do to help get either of these PRs merged!

**Comment by bmuschko** [comment]
@blindpirate I saw that you merged https://github.com/gradle/gradle/pull/1703. Could you please have a look if this PR can simply be closed given that changes are somewhat similar?

**Comment by emlun** [comment]
Yes, the changes suggested here were made in #1703. Thanks!
